Add lens:// protocol handling with a routing mechanism

Nokel81 opened the pull request:
Things that still need to be done:

- [x] Testing on linux
- [x] Testing on windows
- [x] Add markdown docs

Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #1288

Nokel81 committed:
Add lens:// protocol handling with a routing mechanism

- document the methods in an extension guide

- add integration test to make sure that all tested OSes work as
  intended

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on src/extensions/lens-main-extension.ts:
```diff
@@ -16,4 +18,21 @@ export class LensMainExtension extends LensExtension {
+   * @param pathSchema The path schema for the route.
+   * @param handler The function to call when this route has been matched
+   */
+  onProtocol(pathSchema: string, handler: RouteHandler): void {
```
Needs also something to `docs/`.

Nokel81 replied:
Agreed will do.

jakolehm reviewed (commented):
Looks pretty good already 👍 Could we add integration tests for protocol handler?

jim-docker commented on docs/extensions/guides/protocol-handlers.md:
```diff
@@ -0,0 +1,47 @@
+
+Once you have registered a handler it will be called when a user opens a link on their computer.
+The routing mechanism for extensions is quite straight forward.
+For example consider an extension `example-extension` which is publish by the `@mirantis` org.
```
```suggestion
For example consider an extension `example-extension` which is published by the `@mirantis` org.
```

jim-docker commented on docs/extensions/guides/protocol-handlers.md:
```diff
@@ -0,0 +1,47 @@
+Once you have registered a handler it will be called when a user opens a link on their computer.
+The routing mechanism for extensions is quite straight forward.
+For example consider an extension `example-extension` which is publish by the `@mirantis` org.
+If it were to register a handler with `"/display/:type"` as its corresponding then we would match the following URI like this:
```
```suggestion
If it were to register a handler with `"/display/:type"` as its corresponding link then we would match the following URI like this:
```

jim-docker commented on src/common/protocol-handler/router.ts:
```diff
@@ -0,0 +1,205 @@
+    return false;
+  }
+
+  return false;
```
```suggestion
  return true;
```

jim-docker commented on src/extensions/lens-main-extension.ts:
```diff
@@ -16,4 +18,15 @@ export class LensMainExtension extends LensExtension {
+    const lprm = LensProtocolRouterMain.getInstance<LensProtocolRouterMain>();
+
+    lprm.extensionOn(this.name, pathSchema, handler);
+  }
```
Can this be defined in `LensExtension` class instead? (and not duplicated in `LensRendererExtension`)

Nokel81 replied:
No, because the renderer needs to call the `LensProtocolRouterRenderer`'s method instead of the `LensProtocolRouterMain`'s method like here.

jim-docker commented on src/main/protocol-handler/__test__/router.test.ts:
```diff
@@ -0,0 +1,155 @@
+    expect(called).toBe("foo");
+  });
+
+  it("should call most exact handler for an extensions", async () => {
```
```suggestion
  it("should call most exact handler for an extension", async () => {
```

stefcameron commented on src/common/protocol-handler/router.ts:
```diff
@@ -0,0 +1,205 @@
+ * @param args a deserialized value
+ */
+export function validateDeregisterParams(args: unknown): args is DeregisterParams {
+  if (args == null || typeof args !== "object") {
```
A much safer check would be `if (!args || typeof args !== "object")`

Nokel81 replied:
How is that safer? `typeof args === "object"` implies either args is `{...}` or `null`.

stefcameron replied:
But the preceding `!arg` will catch the `null`, `undefined`, `""`, whatever flasy case.

Note that `typeof [] === 'object'` too, so if you're wanting this to be bulletproof, there's still more to do, but technically, an Array is an object and you can safely get/set any property on it, so I figured that wasn't a concern here.

Technically, `arg == null` should also catch the `undefined` case; it's just so foreign to see a `==` in a JS code base. 😄 Implicit coercion of values on comparison is the source of much trouble...

jim-docker replied:
I agree, I like `!args` better than `args == null`. I don't like using the truthy operator `==`. Whenever I see it I wonder if they meant to use `===`.

Nokel81 replied:
Yeah it isn't a concern to me that `typeof [] === "object"`, also `!args` is still a form of implicit coercion. I'll change it if you guys want though. I just felt that this was more explicit about what I wanted.

stefcameron commented on src/common/protocol-handler/router.ts:
```diff
@@ -0,0 +1,205 @@
+ * @param args a deserialized value
+ */
+export function validateRouteParams(args: unknown): args is RouteParams {
+  if (args == null || typeof args !== "object") {
```
Better as `if (!args || typeof args !== "object")`

Nokel81 committed:
Auto-clean extension handlers on deactivation

- remove handlers when an extension is deactivated or removed

- make sure that the found extension when routing a request is currently
  enabled (as a backup)

- added documentation about the above behaviour to the guide

- tweaked the naming convension so that it is clearer that the router
  uses extension names as not IDs (which currently are folder paths)

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on docs/extensions/guides/protocol-handlers.md:
```diff
@@ -45,3 +45,9 @@ On the other hand, the subpath `"/display/notification"` would be routed to #3.
+
+### Cleaning Up
+
+Currently there is not way to remove a protocol handler once it has been registered.
```
...no way to remove...

Why say "currently", are we planning to add a way?

I'm not sure this will fly, there may be use cases for removing a handler while an extension is active. Why don't you want to provide `removeProtocolHandlers()` method in the extension classes as in your hack-week project?

What about a handler being removed by calling `onProtocolRequest()` with the desired pathSchema and null for the handler?

Nokel81 replied:
`removeProtocolHandlers` removes *all* handlers. Yes calling `onProtocolRequest` would remove the handler. I didn't add a removal mechanism because I didn't think that was in the current design (though TBH I was never really given a design, I just asked some questions).

The alternative that I made for the old `removeProtocolHandlers` was to just call it directly for the extension developer.

stefcameron replied:
+1 from me! I think the ability to remove is something we need to provide. Can you add an event handler without ever being able to remove it...? If an extension wanted to stop listening to protocol requests, they'd have to implement some sort of mechanism to know to _ignore_ the requests. That's not very nice.

My vote would be for not overloading the meaning of `onProtocolRequest()` to also handle the removal. I think we should make removal explicit with a "remove" method.

Actually, it feels very much like adding/removing event handlers. The most intuitive thing there would be to rename `onProtocolRequest()` (which, frankly, sounds like a _handler_ itself, not the method you call to _add_ a handler) to `addProtocolHandler()`, and then provide a `removeProtocolHandler()` method.

Nokel81 replied:
Sounds good I can do that. Though I still think that we should remove the handlers when an extension is deactivated.

stefcameron replied:
`onProtocolRequest` is more the name you'd give to a React prop on a component that provides a protocol event. But there, you assign your handler to it, and if you don't want to handle the event anymore, you assign `null/undefined` to it.

In this case, it's a method, so I think a different naming scheme would be more appropriate...

Here's what it looks like in my extension, for example:
https://github.com/Mirantis/lens-extension-cc/blob/master/src/renderer.tsx#L128-L131

It would feel more natural to me if it were:
```javascript
      this.addProtocolHandler(
        `/${EXT_EVENT_ACTIVATE_CLUSTER}`,
        this.handleProtocolActivateCluster
      );
      ...
      this.removeProtocolHandler(this.handleProtocolActivateCluster);
```

Oh, and I guess I'm advocating for a way to remove __individual__ handlers, not just all handlers (the hackweek branch just gave us `removeProtocolHandlers()` -- like all or nothing).

Nokel81 replied:
I know you are asking for removing individual handlers. However, I won't support removing handler by the function because I am sure that it might come up that an extension wants to use the same handler for multiple routes.

stefcameron replied:
Then the `removeProtocolHandler()` can require the route as well.

Nokel81 committed:
Convert the extension API to use an array for registering handlers

- switch design to execute both main and renderer handlers
  simaltaneously, without any overlap checking

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix unit tests and update docs

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
reintroduce renderer routing

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix unit tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

stefcameron commented on docs/extensions/guides/protocol-handlers.md:
```diff
@@ -0,0 +1,57 @@
+
+## Registering A Protocol Handler
+
+The field `protocolhandlers` exists both on [`LensMainExtension`](extensions/api/classes/lensmainextension/#protocolhandlers) and on [`LensRendererExtension`](extensions/api/classes/lensrendererextension/#protocolhandlers).
```
Isn't it `protocolHandlers`? (capital "H") Looks like it in `src/extensions/lens-extension.ts`

Nokel81 replied:
It is, but the api docs generator makes everything lowercase because URLs are case-insensitive.

stefcameron replied:
But this isn't in a URL. See "The field -> protocolhandlers <- exists both on..."

Nokel81 replied:
True, however it is an "inter-doc" link pointing to a generated markdown file. I figured out what it would be by looking at https://docs.k8slens.dev/latest/extensions/api/classes/lensmainextension/#appmenus which is also all lower case even though if you go to it you will see that the field itself is `appMenus`

stefcameron replied:
🤔 I'm not sure we're looking at the same piece of text. I acknowledge that the _anchor_ or _fragment_ in `lensmainextension/#protocolhandlers` and `lensrendererextension/#protocolhandlers` will be all lowercase, but you're telling me that the property name in plain text that precedes it in `` "The field `protocolhandlers` exists both on" `` also has to be lowercase because that somehow becomes a link? I don't mean to press the issue; just seems strange, and looks like a typo, but if that really is the case, then okidoki. 🤷‍♂️

Nokel81 replied:
oh, sorry my bad. I was too narrow focused on the fragment.

stefcameron commented on src/extensions/registries/protocol-handler-registry.ts:
```diff
@@ -0,0 +1,10 @@
+ * a handler to a specific path or dynamic path.
+ */
+export interface ProtocolHandlerRegistration {
+  pathSchema: string;
```
Seems a bit over the top to name it "pathSchema". How about just "path"?

Nokel81 replied:
Why is it over the top? it is a schema in so far as you can describe the shape of a URL.

For instance this is one that I use internally:

```
/:extPubName(\@[A-Za-z0-9_]+)?/:extName`
```

which defines a two part segment where the first must match the regex `/@[A-Za-z0-9_]+/` but is also optional.

stefcameron replied:
I get why it's not _just_ a plain path and there's a schema to it, but react-router, for example, still just calls it a "path": `<Route path="/users/profile/:id">` Just seems simpler in terms of naming. I thought I'd point that out as a comparison.

stefcameron reviewed (commented):
@Nokel81 I'm trying to run Lens from this branch against my extension to test. I've made necessary updates to my extension for the new API. I can't run my extension, though, because of a mobx "max callstack side exceeded" extension the second Lens tries to load my extension:
```
Uncaught (in promise) RangeError: Maximum call stack size exceeded
    at isObservableObject (mobx.module.js:4409)
    at _isObservable (mobx.module.js:2550)
    at isObservable (mobx.module.js:2560)
    at deepEnhancer (mobx.module.js:401)
    at new ObservableValue (mobx.module.js:1031)
    at ObservableObjectAdministration../node_modules/mobx/lib/mobx.module.js.ObservableObjectAdministration.addObservableProp (mobx.module.js:4209)
    at mobx.module.js:453
    at decorate (mobx.module.js:363)
    at decoratorFactory (mobx.module.js:384)
    at observableDecorator (mobx.module.js:463)
```
I can't figure out what this is coming from other than somewhere inside mobx. My extension works fine with Lens 4.0.8. Any ideas?

Nokel81 committed:
fix integration tests and markdown

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on electron-builder.yml:
```diff
@@ -0,0 +1,2 @@
+fileAssociations:
+  - lens
```
Need to verify if this is enough for `snap` build.

Nokel81 replied:
Is `snap` linux or windows? There is an integration test (that should be run on all platforms) that verifies that the application in indeed getting the `open-url` event on all platforms. But yes more manual testing would be good.

jakolehm replied:
Snap as in https://snapcraft.io/. Snap apps are sandboxed and might require additional configuration for host level protocol handlers (we can create a separate issue about verifying this).

Nokel81 replied:
Okay sounds good.

jakolehm commented on src/extensions/extension-loader.ts:
```diff
@@ -52,6 +52,20 @@ export class ExtensionLoader {
+      res.set(val.manifest.name, val);
+    }
+
+    return res;
```
```suggestion
    const extensions = new Map();

    for (const [, val] of this.instances.toJS()) {
      if (val.isBundled) {
        continue;
      }

      extensions.set(val.manifest.name, val);
    }

    return extensions;
```

Nokel81 replied:
Sounds good

jakolehm commented on src/extensions/lens-renderer-extension.ts:
```diff
@@ -29,8 +29,7 @@ export class LensRendererExtension extends LensExtension {
-  // eslint-disable-next-line unused-imports/no-unused-vars-ts
   async isEnabledForCluster(cluster: Cluster): Promise<Boolean> {
-    return true;
+    return (void cluster) || true;
```
Is this just to make `eslint` happy?

Nokel81 replied:
yes

jakolehm commented on src/main/index.ts:
```diff
@@ -34,126 +35,132 @@ let clusterManager: ClusterManager;
+
+    return; // skip exit to make tray work, to quit go to app's global menu or tray's menu
+  })
+  .on("open-url", (event, rawUrl) => {
```
This kind of change to coding style makes huge diff which is really hard to review. Maybe keep the old style just to make the diff smaller?

Nokel81 replied:
Sounds good, I can make a small PR later

jakolehm commented on src/main/protocol-handler/router.ts:
```diff
@@ -0,0 +1,103 @@
+      logger.info(`${proto.LensProtocolRouter.LoggingPrefix}: routing ${url.toString()}`);
+
+      switch (url.host) {
+        case "internal":
```
Does this mean non-extension protocol URLs are like `lens://internal`? Maybe consider more "friendly" name because these links might also be published / used from outside of the app. One candidate is that we route everything except `extensions` to internal.

Nokel81 replied:
do you mean ignoring the `host` of the URI if it is not `extension` or have some way to route based on `host` as well as `pathname`? Or could we change this to something like `app`? The non-extension protocol URIs are `lens://app`.

jakolehm replied:
I meant route based on `host` + `pathname` but if that's complicated then `lens://app` sounds also good.

Nokel81 replied:
It is more complicated since `host` follow different rules from `pathname` and it doesn't gain us anything in terms of flexibility (imo).

jakolehm replied:
Ok, let's use `lens://app`.

stefcameron replied:
👍

Nokel81 committed:
change internal hostname to app, revert spacing

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
resolve pr comment

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix unit and integration tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
testing handling on linux

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix package.json

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Commit:
get protocol handling working for linux

Signed-off-by: Sebastian Malton <smalton@mirantis.com>

Commit:
fix lint

Signed-off-by: Sebastian Malton <smalton@mirantis.com>

jakolehm commented on integration/__tests__/app.tests.ts:
```diff
@@ -87,6 +113,19 @@ describe("Lens integration tests", () => {
+
+        console.log(res);
+
+        await utils.waitForLogsToContain(app, {
```
This might not be a good idea. Infinite loop behind this util will cause problems. Could we have few (one?) real handlers that we can test here. For example `lens://app/preferences` which would open Preferences view.

Nokel81 replied:
How is the infinite loop any worse then for `app.client.waitUntilTextExists`? Sure I can add a single real handler too.

jakolehm replied:
I see lot's of integration failures that leave an app instance behind (we can really have just one instance running)... I'm suspecting it's because this loop never dies.

Nokel81 replied:
having commented out the tests that have this code I still see those instances behind so I doubt that it has to do with that code (plus I have added a fail fast technique to stop the loop when the test ends (from timeout))

jakolehm commented on package.json:
```diff
@@ -210,8 +217,10 @@
+    "moment": "^2.26.0",
     "node-pty": "^0.9.0",
     "npm": "^6.14.8",
+    "open": "^7.3.1",
```
Should this be in `devDependencies`?

Nokel81 replied:
Good idea

Nokel81 committed:
change open to be a dev dep

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
stop getting logs on test timeout

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix integration tests locally

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on docs/extensions/guides/protocol-handlers.md:
```diff
@@ -0,0 +1,57 @@
+# Lens Protocol Handlers
```
This is not really a guide (there's no code here showing how to add a handler to a main or renderer extension class). Could be done in another PR though.

Nokel81 replied:
I have added a code example. Thanks for catching this (as I had not yet exported the types at all)

jim-docker commented on extensions/example-extension/package-lock.json:
```diff
@@ -626,7 +626,684 @@
     "@k8slens/extensions": {
       "version": "file:../../src/extensions/npm/extensions",
-      "dev": true
+      "dev": true,
```
Is all this necessary for protocol handlers?

Nokel81 replied:
Nope, thanks for the catch

jim-docker commented on src/common/protocol-handler/router.ts:
```diff
@@ -0,0 +1,253 @@
+export const ProtocolHandlerExtension= `${ProtocolHandlerIpcPrefix}:extension`;
+
+/**
+ * These two names are long and cubersome by design so as to decrease the chances
```
cumbersome (sp)

Nokel81 committed:
add waiting for renderer and extensions to load before broadcasting handlers

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
revert bad rebase

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
killall during teardown

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/main/protocol-handler/router.ts:
```diff
@@ -0,0 +1,112 @@
+   * is not found while routing to the extensions
+   * @param handler A function that tries to find an extension
+   */
+  public addMissingExtensionHandler(handler: FallbackHandler): void {
```
I don't understand the "missing extension handlers" concept, and this function is not called in the existing code base. When would this be called and by whom?

Nokel81 replied:
This is something that existed from the hackweek, it is true that `addMissingExtensioHandler` isn't called but the list of handlers is.

This is there to facilitate Lens trying to figure out why an extension doesn't have any handlers. Such as:
- it isn't enabled
- it exists (in some extension library) but isn't installed

I made the decision not to include any of those in this PR as I felt it was out of scope of those. But I left the mechanism in as it was a good constraint for me while making sure that the internal design and implementation went smoothly.

jim-docker commented on src/main/window-manager.ts:
```diff
@@ -15,6 +15,9 @@ export class WindowManager extends Singleton {
   protected disposers: Record<string, Function> = {};
 
+  @observable mainViewInitiallyLoaded = false;
+  whenLoaded = when(() => this.mainViewInitiallyLoaded);
```
Is this related to protocol handlers? Maybe you can add a comment here about what impact this has (i.e. what happens when `this.mainViewInitiallyLoaded` gets set to true)

Nokel81 replied:
Yup, this is needed to make sure that the `renderer` has loaded before broadcasting to the renderer because before this is set the renderer is not guaranteed to be in existence.

Sure I can add that as a comment.

jim-docker reviewed (commented):
Seems to match an extension that doesn't exist. For
```
open "lens://extension/nosuchextension/nosuchpath"
```
the logs show:
```
[0] info: [PROTOCOL ROUTER]: routing lens://extension/nosuchextension/nosuchpath
[0] info: [PROTOCOL ROUTER]: Extension nosuchextension matched, but not installed
```

Nokel81 committed:
improve docs, export types for extensions, skip integration tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
add install docs for linux

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix unit tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on docs/extensions/guides/protocol-handlers.md:
````diff
@@ -15,6 +15,23 @@ Handlers will be run in both `main` and `renderer` in parallel with no synchroni
+  ]
+}
+```
+
````
You could also mention the specific URI that would result in calling rootHandler, like using
```
open "lens://extension/example_extension/
```
giving an end-to-end example

jim-docker commented on docs/getting-started/README.md:
```diff
@@ -28,6 +28,28 @@ Review the [System Requirements](/supporting/requirements/) to check if your com
 
 See the [Download Lens](https://github.com/lensapp/lens/releases) page for a complete list of available installation options.
 
+After installing Lens manually (not using a package manager file such as `.deb` or `.rpm`) the following will need to be done in allow protocol handling.
```
```suggestion
After installing Lens manually (not using a package manager file such as `.deb` or `.rpm`) the following will need to be done to allow protocol handling.
```

jim-docker commented on integration/helpers/utils.ts:
```diff
@@ -110,6 +110,7 @@ export async function tearDown(app: Application) {
 
   try {
     process.kill(pid, "SIGKILL");
+    spawnSync("killall lens");
```
Will this kill my Lens 4.0.8 that I might have open while running the integration tests?

Nokel81 replied:
Yes, thanks for reminding me. I added this for testing purposes and should remove it.

Nokel81 committed:
improve docs

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' of github.com:lensapp/lens into feature/protocol-handler

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
remove killall from integration tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix bad merge

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
create dir if necessary (for testing)

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
remove linux based protocol handling setup for testing

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' of github.com:lensapp/lens into feature/protocol-handler

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' of github.com:lensapp/lens into feature/protocol-handler

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Commit:
remove unneccessary shorter timeouts on before* handles in integration tests

Signed-off-by: Sebastian Malton <smalton@mirantis.com>

Commit:
move remove config to make targets

Signed-off-by: Sebastian Malton <smalton@mirantis.com>

Commit:
fix lint

Signed-off-by: Sebastian Malton <smalton@mirantis.com>

Nokel81 committed:
fail tests in before and after handles

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
switch to event emitting renderer being ready

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Add logging and fix renderer:loaded send to main

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' of github.com:lensapp/lens into feature/protocol-handler

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
revert integration test code

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
ignore removeing old config

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/main/app-updater.ts:
```diff
@@ -26,7 +27,7 @@ function handleAutoUpdateBackChannel(event: Electron.IpcMainEvent, ...[arg]: Upd
  * @param interval milliseconds between interval to check on, defaults to 24h
  */
-export function startUpdateChecking(interval = 1000 * 60 * 60 * 24): void {
+export const startUpdateChecking = once(function (interval = 1000 * 60 * 60 * 24): void {
```
Does this fix an unrelated issue? Or just refactoring while accommodating protocol handling?

Nokel81 replied:
It was fixing an issue due to the changes related to `src/main/index.ts:158`. Namely, if `"renderer:loaded"` gets called multiple times `startUpdateChecking` shouldn't be called on the subsequent events.

jim-docker replied:
I figured, but is it still a bug otherwise? If so, shouldn't it really be in its own PR?

Nokel81 replied:
I figured that I should fix it in this PR since it was in this PR that I changed how the `renderer` lets `main` know it has been loaded.

jim-docker commented on src/main/index.ts:
```diff
@@ -37,30 +38,54 @@ let windowManager: WindowManager;
 
 app.setName(appName);
 
+logger.info("📟 Setting as Lens as protocol client for lens://");
```
```suggestion
logger.info("📟 Setting Lens as protocol client for lens://");
```

Nokel81 replied:
Good catch, though I will do that in a follow up PR so as to not need new approves.

jakolehm reviewed (approved):
LGTM. Let's keep improving this in follow-up PRs.
